controller/program/tests/tests/rebalance/chain.rs

```diff
@@ -0,0 +1,423 @@
+) {
+    assert_eq!(result.program_result, ProgramResult::Success);
+
+    let pool_pk = Pubkey::new_from_array(POOL_STATE_ID);
+    let pool_bef = accounts_before
+        .iter()
+        .find(|(pk, _)| *pk == pool_pk)
+        .map(|(_, acc)| acc)
+        .expect("pool before");
+    let pool_aft = result
+        .resulting_accounts
+        .iter()
+        .find(|(pk, _)| *pk == pool_pk)
+        .map(|(_, acc)| acc)
+        .expect("pool after");
+
+    let pool_state_bef = PoolStatePacked::of_acc_data(&pool_bef.data)
+        .expect("pool before")
+        .into_pool_state();
+    let pool_state_aft = PoolStatePacked::of_acc_data(&pool_aft.data)
+        .expect("pool after")
+        .into_pool_state();
+
+    assert_eq!(pool_state_bef.is_rebalancing, 0);
+    assert_eq!(pool_state_aft.is_rebalancing, 0);
+    assert!(pool_state_aft.total_sol_value >= old_total_sol_value,);
+
+    // Assert RebalanceRecord is closed
+    let rr_pk = Pubkey::new_from_array(REBALANCE_RECORD_ID);
+    let rr_after = result
+        .resulting_accounts
+        .iter()
+        .find(|(pk, _)| *pk == rr_pk);
+    assert!(rr_after.is_none() || rr_after.unwrap().1.lamports == 0);
+
+    assert_balanced(accounts_before, &result.resulting_accounts);
+}
+
+#[test]
+fn rebalance_transaction_success() {
+    let amount = 100_000;
+    let (accounts_before, result, old_total_sol_value) = execute_rebalance_transaction(amount);
+
+    assert_rebalance_transaction_success(&accounts_before, &result, old_total_sol_value);
+
+    let mut result_for_check = result.clone();
+
+    if let Some(token_prog_before) = accounts_before
+        .iter()
+        .find(|(pk, _)| pk.to_bytes() == TOKENKEG_PROGRAM)
+    {
+        if let Some(token_prog_result) = result_for_check
+            .resulting_accounts
+            .iter_mut()
+            .find(|(pk, _)| pk.to_bytes() == TOKENKEG_PROGRAM)
+        {
+            token_prog_result.1 = token_prog_before.1.clone();
+        }
+    }
+
+    if let Some(instructions_sysvar_before) = accounts_before
+        .iter()
+        .find(|(pk, _)| pk.to_bytes() == INSTRUCTIONS_SYSVAR_ID)
+    {
+        if let Some(instructions_sysvar_result) = result_for_check
+            .resulting_accounts
+            .iter_mut()
+            .find(|(pk, _)| pk.to_bytes() == INSTRUCTIONS_SYSVAR_ID)
+        {
+            instructions_sysvar_result.1 = instructions_sysvar_before.1.clone();
+        }
+    }
+
+    // Assert all accounts are rent-exempt after transaction
+    SVM.with(|svm| {
+        assert!(result_for_check.run_checks(&[Check::all_rent_exempt()], &svm.config, svm));
+    });
+}
+
+proptest! {
+    #[test]
+    fn rebalance_transaction_various_amounts_any(
+        amount in 1_000u64..=1_000_000
+    ) {
+        let (accounts_before, result, old_total_sol_value) = execute_rebalance_transaction(amount);
+
+        assert_rebalance_transaction_success(&accounts_before, &result, old_total_sol_value);
+    }
+}
```

Comment:
https://github.com/igneous-labs/inf-1.5/pull/89/files#r2495092935

again, you want to be testing the failure cases. It's loss of funds disasta if someone is able to StartRebalance without a corresponding EndRebalance